primitives/s2s/src/lib.rs

```diff
@@ -0,0 +1,168 @@
+// This file is part of Darwinia.
+//
+// Copyright (C) 2018-2021 Darwinia Network
+// SPDX-License-Identifier: GPL-3.0
+//
+// Darwinia is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Darwinia is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Darwinia. If not, see <https://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+// --- paritytech ---
+use sp_core::H160;
+use sp_std::{vec, vec::Vec};
+// --- darwinia-network ---
+use codec::{Decode, Encode};
+use dp_asset::token::Token;
+use dp_contract::mapping_token_factory::s2s::S2sRemoteUnlockInfo;
+
+/// The parameters box for the pallet runtime call.
+#[derive(Encode, Decode, Debug, PartialEq, Eq, Clone)]
+pub enum CallParams<AccountId> {
+	#[codec(index = 0)]
+	S2sIssuingPalletRegisterFromRemote(Token),
+	#[codec(index = 1)]
+	S2sIssuingPalletIssueFromRemote(Token, H160),
+	#[codec(index = 2)]
+	S2sBackingPalletUnlockFromRemote(AccountId, S2sRemoteUnlockInfo),
+}
+
+/// Creating a concrete message payload which would be relay to target chain.
+pub trait PayloadCreate<AccountId, MessagePayload>
+where
+	AccountId: Encode + Decode,
+{
+	fn encode_call(
+		pallet_index: u8,
+		call_params: CallParams<AccountId>,
+	) -> Result<Vec<u8>, &'static str> {
+		let mut encoded = vec![pallet_index];
+		encoded.append(&mut call_params.encode());
```

Comment:
this method to encode the remote call has a problem that we can't know exactly which method to encode. Scale or others? And we know we use the scale encode. But if the implementation of the scale encode changed(version upgrade), then it can't update with it.